lib/prometheus.js

```diff
@@ -0,0 +1,135 @@
+const PromClient = require('prom-client');
+const uuid = require("uuid");
+const debug = require('debug')('plugin:publish-metrics:prometheus');
+
+const COUNTERS_STATS = 'counters', // counters stats
+  RATES_STATES = 'rates', // rates stats
+  SUMMARIES_STATES = 'summaries'; // summaries stats includes errors
+
+class PrometheusReporter {
+
+  constructor(config, events) {
+    this.jobUUID = uuid.v4();
+    this.config = Object.assign({
+      tags: [],
+      prefix: 'artillery'
+    }, config)
+
+    this.prometheusOpts = {
+      pushgatewayUrl: config.pushgateway,
+    };
+
+    debug('ensuring pushgatewayUrl is configured');
+    if (!this.prometheusOpts.pushgatewayUrl) {
+      console.error(`the prometheus [pushgateway] url was not specified`);
+    }
+
+    debug('setting default labels');
+    PromClient.register.setDefaultLabels(this.tagsToLabels(this.config.tags));
+
+    this.registerMetrics(this.config.prefix)
+
+    debug('creating pushgateway client using url: %s', this.prometheusOpts.pushgatewayUrl);
+    this.pushgateway = new PromClient.Pushgateway(this.prometheusOpts.pushgatewayUrl);
```

Comment:
Might be a good idea to set `http.Agent` options here - I'm thinking `keepAlive` specifically. Otherwise in tests with hundreds or thousands of workers we'd potentially be opening up lots of TCP connections to Pushgateway every 10 seconds.